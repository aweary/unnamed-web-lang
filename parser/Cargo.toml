[package]
name = "parser"
version = "0.1.0"
authors = ["Brandon Dail <brandondail@fb.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
graphql-parser = "0.2.3"
salsa = "0.14.1"
lexer = { path = "../lexer" }
syntax = { path = "../syntax" }
diagnostics = { path = "../diagnostics" }
source = { path = "../source" }
insta = { version = "0.16", features = ["glob"] }
log = "0.4.8"