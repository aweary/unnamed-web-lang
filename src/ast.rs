#[derive(Debug)]
pub struct Param {
    name: String,
    kind: String,
}

impl Param {
    pub fn new(name: String, kind: String) -> Self {
        Param { name, kind }
    }
}

/**
 * The definition of a component. Similar to a function definition, but with
 * stricter semantics and requirements.
 *
 * @example
 * component Button(label: string) {
 *   <button>{label}</button>
 * }
 *
 */
#[derive(Debug)]
pub struct ComponentDefinition {
    // Name of the component
    name: String,
    // List of its parameters, which must by typed
    params: Option<Vec<Param>>,
    // The content of the component
    body: Vec<Expression>,
}

impl ComponentDefinition {
    pub fn new(name: String, params: Option<Vec<Param>>, body: Vec<Expression>) -> Self {
        ComponentDefinition { name, params, body }
    }
}

#[derive(Debug)]
pub struct TypeDefinition {
    name: String,
}

impl TypeDefinition {
    pub fn new(name: String) -> Self {
        TypeDefinition { name }
    }
}

#[derive(Debug)]
pub struct Identifier {
    name: String,
}

impl Identifier {
    pub fn new(name: String) -> Self {
        Identifier { name }
    }
}

#[derive(Debug)]
pub struct JSXOpeningElement {
    tag_name: String,
    attributes: Option<Vec<JSXAttribute>>,
    self_closing: bool,
}

impl JSXOpeningElement {
    pub fn new(
        tag_name: String,
        attributes: Option<Vec<JSXAttribute>>,
        self_closing: bool,
    ) -> Self {
        JSXOpeningElement {
            tag_name,
            attributes,
            self_closing,
        }
    }
}

#[derive(Debug)]
pub struct JSXClosingElement {
    tag_name: String,
}

impl JSXClosingElement {
    pub fn new(tag_name: String) -> Self {
        JSXClosingElement { tag_name }
    }
}

#[derive(Debug)]
pub struct JSXAttribute {
    name: String,
    value: JSXAttributeValue,
}

impl JSXAttribute {
    pub fn new(name: String, value: JSXAttributeValue) -> Self {
        JSXAttribute { name, value }
    }
}

#[derive(Debug)]
pub struct JSXElement {
    opening_element: JSXOpeningElement,
    closing_element: Option<JSXClosingElement>,
    children: Option<Vec<Box<JSXChildren>>>,
}

impl JSXElement {
    pub fn new(
        opening_element: JSXOpeningElement,
        closing_element: Option<JSXClosingElement>,
        children: Option<Vec<Box<JSXChildren>>>,
    ) -> Self {
        JSXElement {
            opening_element,
            closing_element,
            children,
        }
    }
}

#[derive(Debug)]
pub enum Expression {
    NumericLiteral(i64),
    StringLiteral(String),
    Identifer(Identifier),
    JSXElement(JSXElement),
}

#[derive(Debug)]
pub enum JSXAttributeValue {
    StringLiteral(String),
    Expression(Expression),
}

#[derive(Debug)]
pub enum JSXChildren {
    JSXText(String),
    JSXExpressionContainer(Expression),
    JSXElement(JSXElement),
}

// pub enum Expression {
//     JSXElement,
//     JSXOpeningElement,
//     JSXClosingElement,
//     JSXAttribute,
// }

#[derive(Debug)]
pub enum Node {
    ComponentDefinition(ComponentDefinition),
    TypeDefinition(TypeDefinition),
    // Identifier,
    // OpeningElement,
    // ClosingElement,
    Text,
    // ExpressionStatement,
    // StateCellDeclaration,
}
