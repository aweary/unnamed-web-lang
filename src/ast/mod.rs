mod expr;
mod stmt;

use crate::symbol::Symbol;
use crate::typecheck::TKind;

pub use expr::*;
pub use stmt::*;


// The root node of the AST.
#[derive(Debug)]
pub struct Program {
    // The set of all modules the program requires.
    pub modules: Vec<Module>,
    // Arena that stores all the template references.
}

// An individual module. Currently the only way to define
// a module is with a new file, so this maps directly to the filesystem
// layout of the project.
#[derive(Debug)]
pub struct Module {
    pub stmts: Vec<Decl>,
}

#[derive(Debug, PartialOrd, Ord, PartialEq, Eq, Clone, Copy)]
pub enum Precedence {
    NONE = 0,
    ASSIGNMENT = 1,
    CONDITIONAL = 2,
    SUM = 3,
    PRODUCT = 4,
    COMPARE = 5,
    PREFIX = 6,
}






