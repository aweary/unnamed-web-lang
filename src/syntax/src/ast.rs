use serde::{Deserialize, Serialize};

use crate::symbol::Symbol;
use codespan::Span;

pub use crate::ty::{LiteralTy, Ty};
use diagnostics::ParseResult as Result;

use std::fmt::{Debug, Error, Formatter};
use std::path::PathBuf;

const DUMMY_NODE_ID: NodeId = NodeId(0);

pub fn expr(kind: ExprKind, span: Span) -> Result<Expr> {
    Ok(Expr {
        id: DUMMY_NODE_ID,
        kind,
        span,
        ty: None,
    })
}

// TODO move these into symbols crate
#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct Ident {
    pub name: Symbol,
    pub span: Span,
}

impl Ident {
    pub fn to_str(&self) -> &str {
        &self.name.as_str()
    }
}

impl Debug for Ident {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::result::Result<(), Error> {
        write!(f, "{:?}", self.name)
    }
}

// TODO...
#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct NodeId(pub usize);

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Program {
    pub modules: Vec<Mod>,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Mod {
    pub items: Vec<Item>,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Item {
    pub ident: Ident,
    pub id: NodeId,
    pub kind: ItemKind,
    pub span: Span,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum ItemKind {
    /// A statoc item
    ///
    /// e.g., `static STATIC_NUM : number = 42`
    Static(Box<Ty>, Box<Expr>),
    /// A function declaration
    ///
    /// e.g., `fn add(a: number, b: number) : number { ... }
    Fn(FnDef),
    /// An enum definition
    Enum(EnumDef, Option<Generics>),
    /// Type definition for records
    Type(Box<TypeDef>),
    /// Import declaration
    Import(Box<Import>),
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Import {
    pub name: Ident,
    pub span: Span,
    pub path: ImportPath,
}

impl Import {
    pub fn to_path_buf(&self) -> PathBuf {
        self.clone().path.path
    }
}

#[derive(Hash, Eq, PartialEq, Clone, Serialize, Deserialize)]
pub struct ImportPath {
    pub path: PathBuf,
    pub span: Span,
}

impl Debug for ImportPath {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::result::Result<(), Error> {
        write!(f, "{:?}", self.path)
    }
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct TypeDef {
    pub name: Ident,
    pub span: Span,
    pub generics: Option<Generics>,
    pub properties: Vec<TypeProperty>,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct TypeProperty {
    pub name: Ident,
    pub ty: Ty,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum ParamType {
    // No paramters
    Empty,
    // A single paramter
    Single(Param),
    // >=2 paramters
    Multi(Vec<Param>),
}

impl IntoIterator for ParamType {
    type Item = Param;
    type IntoIter = std::vec::IntoIter<Self::Item>;
    fn into_iter(self) -> Self::IntoIter {
        use ParamType::*;
        match self {
            Empty => vec![].into_iter(),
            Single(param) => vec![param].into_iter(),
            Multi(params) => params.into_iter(),
        }
    }
}

/// A function definition
#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct FnDef {
    pub name: Ident,
    pub params: ParamType,
    pub body: Box<Block>,
    pub return_ty: Ty,
    pub is_async: bool,
    pub generics: Option<Generics>,
    pub span: Span,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct EnumDef {
    pub variants: Vec<Variant>,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Variant {
    pub ident: Ident,
    pub id: NodeId,
    pub span: Span,
    // ...
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct Generics {
    // We don't currenly support nested generic types
    pub params: Vec<Ident>,
    // We don't currently support generic functions, but
    // we'll encode this into the AST just in case.
    // ...
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Block {
    pub stmts: Vec<Stmt>,
    pub id: NodeId,
    pub span: Span,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Stmt {
    pub id: NodeId,
    pub kind: StmtKind,
    pub span: Span,
    pub has_semi: bool,
    // ...
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum StmtKind {
    // A local, let binding
    Local(Box<Local>),
    // An item definition, local to some block
    Item(Box<Item>),
    // Expression statement
    // TODO should we differentiate expressions with or without semicolons?
    Expr(Box<Expr>),
    // A while loop
    While(Box<Expr>, Box<Block>),
    // Return statement
    Return(Box<Expr>),
    // Try/catch statement
    TryCatch(Box<Block>, Option<LocalPattern>, Box<Block>),
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct LocalObjectProperty {
    pub key: Ident,
    pub value: LocalPattern,
    pub span: Span,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum LocalPattern {
    Ident(Ident, Span),
    Object(Vec<LocalObjectProperty>, Span),
    List(Vec<Ident>, Span),
}

impl Into<Symbol> for LocalPattern {
    fn into(self) -> Symbol {
        match self {
            LocalPattern::Ident(ident, _) => ident.name,
            _ => todo!("Cannot make destructured LocalPattern into Symbol"),
        }
    }
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Local {
    pub id: NodeId,
    pub name: LocalPattern,
    pub ty: Option<Box<Ty>>,
    // Optional initializing expression.
    pub init: Option<Box<Expr>>,
    pub span: Span,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct FnDecl {
    pub params: Vec<Param>,
    pub output: Box<Option<Ty>>,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct FnHeader {
    pub is_async: bool,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Param {
    pub local: LocalPattern,
    pub ty: Ty,
    pub id: NodeId,
    pub span: Span,
    // ...
}

impl Param {
    pub fn name(&self) -> Symbol {
        match &self.local {
            LocalPattern::Ident(ident, _) => ident.name.clone(),
            _ => todo!("Cant get name for destructure yet"),
        }
    }
}

// #[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
// pub struct Ty {
//     // TODO do we need a name?
//     pub name: Ident,
//     pub id: NodeId,
//     pub kind: TyKind,
//     pub span: Span,
//     pub generics: Option<Generics>,
// }

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum TyKind {
    /// Placeholder
    Empty,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Expr {
    pub id: NodeId,
    pub kind: ExprKind,
    pub span: Span,
    pub ty: Option<Ty>,
    // ...
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum ExprKind {
    Array(Vec<Expr>),
    Object(Vec<(Ident, Expr)>),
    Tuple(Vec<Expr>),
    /// Block expression
    Block(Box<Block>),
    /// A binary operation
    Binary(BinOp, Box<Expr>, Box<Expr>),
    /// A unary operation
    Unary(UnOp, Box<Expr>),
    /// Conditional expression, e.g., ternary
    Cond(Box<Expr>, Box<Expr>, Box<Expr>),
    /// Call expression
    Call(Box<Expr>, Vec<Expr>),
    /// Assignment expression
    // TODO the left hand side should be a LeftExpr or something
    Assign(AssignOp, Box<Expr>, Box<Expr>),
    /// Object member access
    Member(Box<Expr>, Ident),
    /// Object optional member access,
    OptionalMember(Box<Expr>, Ident),
    /// A literal
    Lit(Lit),
    /// A variable reference
    Reference(Ident),
    /// An `if` block with optional `else` block
    If(IfExpr),
    /// For expression
    For(LocalPattern, Box<Expr>, Box<Block>),
    /// An index operation
    Index(Box<Expr>, Box<Expr>),
    /// A `return` with an optional return expression
    Return(Option<Box<Expr>>),
    /// Template
    Template(Template),
    /// Match
    Match(Box<Expr>, Vec<MatchArm>),
    /// Function expression
    Func(Box<FnDef>),
    // ...
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum Else {
    Block(Box<Block>),
    // TODO this should be IfExpr but our parser types don't
    // work super well for this right now
    If(Box<Expr>),
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct IfExpr {
    pub condition: Box<Expr>,
    pub block: Box<Block>,
    pub alt: Option<Else>,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct MatchArm {}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Template {
    pub id: NodeId,
    pub open: TemplateOpenTag,
    pub close: Option<TemplateCloseTag>,
    pub children: Option<Vec<TemplateChild>>,
    pub span: Span,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum TemplateChild {
    Text(Symbol),
    Template(Box<Template>),
    Expr(Box<Expr>),
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct TemplateOpenTag {
    pub name: Ident,
    pub attrs: Vec<TemplateAttr>,
    pub span: Span,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct TemplateCloseTag {
    pub name: Ident,
    pub span: Span,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct TemplateAttr {
    pub name: Ident,
    pub value: Expr,
    pub span: Span,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct Lit {
    pub span: Span,
    pub kind: LitKind,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum LitKind {
    Bool(Symbol),
    // A number, interned as a symbol. This will be
    // derserialized later on.
    Number(Symbol),
    Str(Symbol),
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum BinOp {
    Equals,
    DblEquals,
    Add,
    Sub,
    Sum,
    Mul,
    Div,
    Mod,
    And,
    Or,
    GreaterThan,
    LessThan,
    Pipeline,
    // ...
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum UnOp {
    Negate,
    Plus,
    Minus,
    Increment,
}

#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum AssignOp {
    Equals,
    PlusEquals,
    MinusEquals,
    MulEquals,
    DivEquals,
}
