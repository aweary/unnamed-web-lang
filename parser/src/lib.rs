#![warn(clippy::all, clippy::pedantic)]
#![allow(clippy::must_use_candidate)]

mod parser;

pub use crate::parser::*;

#[cfg(test)]
mod tests {
    use crate::parser::Parser;
    use insta::assert_json_snapshot;
    use syntax::ast;

    fn parse_module(source: &str) -> ast::Module {
        let mut parser = Parser::new(source);
        parser.parse_module().unwrap()
    }

    #[test]
    fn fixture_function_simple() {
        let input = include_str!("./inputs/function_simple.dan");
        assert_json_snapshot!(parse_module(input));
    }

    #[test]
    fn fixture_control_flow() {
        let input = include_str!("./inputs/control_flow.dan");
        assert_json_snapshot!(parse_module(input));
    }

    #[test]
    fn fixture_trailing_closure_control_flow() {
        let input = include_str!("./inputs/trailing_closures.dan");
        assert_json_snapshot!(parse_module(input));
    }
}
