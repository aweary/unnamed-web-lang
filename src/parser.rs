use ast::*;
use lexer::Lexer;
use token::*;

pub struct Parser {
    pub lexer: Lexer,
    // state: ParserState,
}

impl Parser {
    pub fn module(source: String) -> Self {
        let lexer = Lexer::new(source);
        // let state = ParserState::new();
        Parser { lexer }
    }

    fn peek(&mut self) -> Option<&Token> {
        self.lexer.peek_token()
    }

    pub fn skip_until<F>(&mut self, predicate: F)
    where
        F: Fn(Token) -> bool,
    {
        loop {
            match self.lexer.next_token() {
                Some(token) => {
                    if predicate(token) {
                        return;
                    }
                }
                None => return,
            }
        }
    }

    pub fn skip_assert(&mut self, token: Token) {
        assert_eq!(Some(token), self.lexer.next_token())
    }

    fn read_param_list(&mut self) -> Option<Vec<Param>> {
        match self.lexer.next_token() {
            // If the next token is an opening curly bracket, there
            // are no arguments and we assume this component takes no input
            Some(Token::LCurlyBracket) => None,
            // If the next token is an opening paren, there is
            // an argument list.
            Some(Token::LParen) => {
                let mut params: Vec<Param> = vec![];
                loop {
                    let name = self.read_ident();
                    // TODO this should be an unexpected token error
                    assert_eq!(Some(Token::Colon), self.lexer.next_token());
                    let kind = self.read_ident();
                    params.push(Param::new(name, kind));
                    match self.lexer.next_token() {
                        Some(Token::RParen) => {
                            break;
                            // End of the param list
                        }
                        Some(Token::Comma) => {
                            continue;
                            // There's another paramter
                        }
                        // Anything else is an unexpected token error
                        _ => unreachable!(),
                    }
                }
                Some(params)
            }
            _ => unreachable!("Unexpected token"),
        }
        // Next token should be the opening paren for the argument list
    }

    fn read_ident(&mut self) -> String {
        match self.lexer.next_token() {
            Some(Token::Ident(name)) => name,
            unknown @ _ => unreachable!("Expected identifier, found: {:?}", unknown),
        }
    }

    // Assumes that the current token is Reserved(Component)
    fn read_component_declaration(&mut self) -> Node {
        // Next token should be the identifier for the component definition
        let name = self.read_ident();
        // Read list of all params
        let params = self.read_param_list();
        // Everything from here until the closing bracket is part of the
        // component definition
        let body = self.read_component_body();
        let decl = ComponentDefinition::new(name, params, body);
        Node::ComponentDefinition(decl)
    }

    // A component is composed of a set of expressions/statements.
    fn read_component_body(&mut self) -> Vec<Expression> {
        let mut exprs = vec![];
        loop {
            match self.lexer.next_token() {
                // Statements starting with '<' are assumed to be
                // the beginning of JSX statements.
                Some(Token::LessThan) => {
                    let expr = Expression::JSXElement(self.read_jsx_element());
                    exprs.push(expr);
                    return exprs;
                }
                _ => {
                    continue;
                }
            }
        }
    }

    /**
     * read_expr
     * Reads any expresison, which is something that can evaluate to some value
     */
    fn read_expr(&mut self) -> Expression {
        self.skip_until(|token| token == Token::RCurlyBracket);
        Expression::NumericLiteral(42)
    }

    fn read_jsx_attributes(&mut self) -> Option<Vec<JSXAttribute>> {
        let mut attrs: Vec<JSXAttribute> = vec![];
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let attr = self.read_jsx_attribute();
                    attrs.push(attr);
                }
                Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                    break;
                }
                _ => {
                    unreachable!("Unknown token: {:?}", self.peek());
                }
            }
        }
        println!("read_jsx_attributes: {:?}", attrs);
        if attrs.is_empty() {
            None
        } else {
            Some(attrs)
        }
    }

    fn is_jsx_element_self_closing(&mut self) -> bool {
        match self.peek() {
            Some(&Token::GreaterThan) => {
                self.skip_assert(Token::GreaterThan);
                false
            }
            Some(&Token::ForwardSlash) => {
                self.skip_assert(Token::ForwardSlash);
                self.skip_assert(Token::GreaterThan);
                true
            }
            _ => {
                unreachable!();
            }
        }
    }

    fn read_jsx_element(&mut self) -> JSXElement {
        let tag_name = self.read_ident();
        let attributes = self.read_jsx_attributes();
        let self_closing = self.is_jsx_element_self_closing();
        let opening_element = JSXOpeningElement::new(tag_name.clone(), attributes, self_closing);
        println!("opening_element {:?}", opening_element);
        let (children, closing_element) = if self_closing {
            (None, None)
        } else {
            let children = self.read_jsx_children();
            let closing_element = self.read_jsx_closing_element();
            println!("closing_element {:?}", closing_element);
            println!("children {:?}", children);
            (Some(children), Some(closing_element))
        };
        let element = JSXElement::new(opening_element, closing_element, children);
        println!("element: {:#?}", element);
        element
    }

    fn read_jsx_closing_element(&mut self) -> JSXClosingElement {
        // Closing tag
        // self.skip_assert(Token::LessThan);
        // self.skip_assert(Token::ForwardSlash);
        let closing_tag_name = self.read_ident();
        // TODO parsing error
        // assert_eq!(tag_name, closing_tag_name);
        self.skip_assert(Token::GreaterThan);
        // TODO actually parse the closing tag
        JSXClosingElement::new(closing_tag_name)
    }

    fn read_jsx_text(&mut self) -> Box<JSXChildren> {
        let mut text = String::new();
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let word = self.read_ident();
                    // TODO respect whitespace when lexing instead of this hack
                    text = text + " " + &word;
                }
                Some(&Token::LessThan) | Some(&Token::LCurlyBracket) => {
                    break;
                }
                _ => {
                    unreachable!();
                }
            }
        }
        println!("read_jsx_text: {:?}", text);
        Box::new(JSXChildren::JSXText(text))
    }

    fn read_jsx_children(&mut self) -> Vec<Box<JSXChildren>> {
        let mut children: Vec<Box<JSXChildren>> = vec![];
        loop {
            match self.peek() {
                // JSXText
                Some(&Token::Ident(_)) => {
                    children.push(self.read_jsx_text());
                },
                Some(&Token::LessThan) => {
                    // A '<' token could be the start of another element
                    // or the closing element for the enclosing element
                    self.skip_assert(Token::LessThan);
                    match self.peek() {
                        Some(&Token::Ident(_)) => {
                            // An identifier means the start of a new element
                            let element = self.read_jsx_element();
                            children.push(Box::new(JSXChildren::JSXElement(element)));
                        }
                        Some(&Token::ForwardSlash) => {
                            self.skip_assert(Token::ForwardSlash);
                            return children;
                            // A '/' indicates a closing element
                        }
                        _ => {
                            unreachable!();
                        }
                    }
                }
                _ => break,
            }
        }
        children
        // let child = match token {
        //     // TODO JSX parsers respect the whitespace inside tags. Need to output tokens
        //     Some(Token::Ident(word)) => {
        //         JSXChildren::JSXText(word)
        //     }
        //     Some(Token::LessThan) => {
        //         let element = self.read_jsx_element();
        //         JSXChildren::JSXElement(element)
        //     }
        //     // TODO handle JSXExpressionContainer
        //     token @ _ => {
        //         unreachable!("Unexpected token: {:?}", token);
        //     }
        // };
        // Box::new(child)
    }

    /**
     * Reads the expression part of a JSX attribute.
     */
    fn read_jsx_attribute_value(&mut self) -> JSXAttributeValue {
        // An explicit assignment. The assigned value can
        // either be a string literal or an identifier in curly brackets
        match self.lexer.next_token() {
            Some(Token::String(literal)) => JSXAttributeValue::StringLiteral(literal),
            Some(Token::LCurlyBracket) => {
                let expr = self.read_expr();
                JSXAttributeValue::Expression(expr)
            }
            token @ _ => unreachable!("Unknown token: {:?}", token),
        }
    }

    /**
     * read_jsx_attr
     *
     * A JSX Attribute comes in two forms:
     *
     * 1. An explicit attribute where the value is passed
     *    in inline (<div id="1" />)
     * 2. An implicit attribute where we assume the attribute
     *    name is a reference to some identifier
     *    (<div id>)
     */
    fn read_jsx_attribute(&mut self) -> JSXAttribute {
        let name = self.read_ident();

        let foo = match self.peek() {
            Some(&Token::Assign) => {
                self.skip_assert(Token::Assign);
                let attribute_value = self.read_jsx_attribute_value();
                JSXAttribute::new(name, attribute_value)
            }
            Some(&Token::Ident(_)) | Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                let attr_ident = Identifier::new(name.clone());
                let attr_ident_expr = Expression::Identifer(attr_ident);
                let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
                JSXAttribute::new(name, attr_value)
            }
            _ => {
                unreachable!();
            }
        };
        foo
        // println!("read_jsx_attribute:name {:?}", name);
        // match self.lexer.next_token() {
        //     // Explicit assignment
        //     Some(Token::Assign) => {
        //         let attr_value = self.read_jsx_attribute_value();
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     // Shorthand form
        //     Some(Token::Ident(_)) | Some(Token::ForwardSlash) | Some(Token::GreaterThan) => {
        //         println!("shorthand form, creating the identifier and stuff");
        //         let attr_ident = Identifier::new(name.clone());
        //         let attr_ident_expr = Expression::Identifer(attr_ident);
        //         let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     token @ _ => unreachable!("Unknown character: {:?}", token),
        // }
    }

    fn read_type_def(&mut self) -> Node {
        // Next token is the identifier for the type definition.
        let name = self.read_ident();
        let def = TypeDefinition::new(name);
        // Just read everything until we reach the end for now...
        loop {
            if self.lexer.next_token() == Some(Token::RCurlyBracket) {
                break;
            }
        }
        Node::TypeDefinition(def)
    }

    /**
     * Inside a module, top level nodes are currently restricted to component
     * and function definitions. This means you cannot have other expressions
     * that are executed when the module is evalulated. All components and
     * functions must be lazily evaluated.
     */
    pub fn next_node(&mut self) -> Node {
        let token = self.lexer.next_token().unwrap();
        match token {
            Token::Reserved(ReservedWord::Component) => self.read_component_declaration(),
            Token::Reserved(ReservedWord::Type) => self.read_type_def(),
            _ => Node::Text,
        }
    }
}
