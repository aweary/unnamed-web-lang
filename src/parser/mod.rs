mod decl;
mod expr;
mod stmt;
mod context;

use decl::DeclParser;
// use expr::ExprParser;
// use stmt::StmtParser;
pub use context::{ParsingContext, ExprContext};


use crate::error::ParseError;
use crate::lexer::Lexer;
use crate::result::Result;
use crate::ast::{Module, Program};
use crate::symbol::Symbol;
use crate::token::{Token, TokenKind};
use std::rc::Rc;
use std::cell::RefCell;

// temporary use so that the Rust compiler checks this module
// use crate::ir::*;

pub fn parse_program<'a>(source: &'a str, mut ctx: &'a mut ParsingContext) -> Result<Program> {
    let mut parser = Parser::new(&source, &mut ctx);
    parser.parse()
}

pub struct Parser<'a> {
    lexer: Lexer<'a>,
    ctx: &'a mut ParsingContext,
}

impl<'a> Parser<'a> {
    pub fn new(source: &'a str, ctx: &'a mut ParsingContext) -> Self {
        let lexer = Lexer::new(&source);
        Parser { lexer, ctx }
    }

    fn next_token(&mut self) -> Result<Token> {
        match self.lexer.next_token(&mut self.ctx) {
            Ok(token) => {
                Ok(token)
            }
            Err(_) => Err(ParseError::LexError),
        }
    }

    fn peek_token(&mut self) -> Result<&Token> {
        match self.lexer.peek_token(&mut self.ctx) {
            Ok(token) => Ok(token),
            Err(_) => Err(ParseError::LexError),
        }
    }

    fn eat(&mut self, kind: TokenKind) -> Result<bool> {
        if self.peek_token()?.kind == kind {
            self.expect(kind)?;
            Ok(true)
        } else {
            Ok(false)
        }
    }

    fn expect(&mut self, kind: TokenKind) -> Result<Token> {
        let token = self.next_token()?;
        if token.kind != kind {
            Err(ParseError::UnexpectedToken(token))
        } else {
            Ok(token)
        }
    }

    fn ident(&mut self) -> Result<Symbol> {
        let token = self.next_token()?;
        match token.kind {
            TokenKind::Ident(sym) => Ok(sym),
            _ => Err(ParseError::UnexpectedToken(token)),
        }
    }

    /**
     * This is currently an alias to `ident`, but a type
     * identifier has different syntax. For example, polymorphic
     * types have type parameters.
     */
    fn type_ident(&mut self) -> Result<Symbol> {
        self.ident()
    }

    pub fn parse(&mut self) -> Result<Program> {
        // top-level statements
        let stmts = self.decl_list()?;
        let module = Module { stmts };
        let modules = vec![module];
        let program = Program { modules };
        Ok(program)
    }
}
