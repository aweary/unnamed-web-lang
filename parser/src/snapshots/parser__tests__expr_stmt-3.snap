---
source: src/lib.rs
expression: parse_statement(input)
---
Stmt {
    id: NodeId(
        0,
    ),
    kind: Expr(
        Expr {
            id: NodeId(
                0,
            ),
            kind: Lit(
                Lit {
                    span: Span {
                        start: ByteIndex(0),
                        end: ByteIndex(7),
                    },
                    kind: Str(
                        "Hello",
                    ),
                },
            ),
            span: Span {
                start: ByteIndex(0),
                end: ByteIndex(7),
            },
            ty: None,
        },
    ),
    span: Span {
        start: ByteIndex(0),
        end: ByteIndex(7),
    },
    has_semi: false,
}
